### DISCUSSION TITLE
A good way to process an array?

### DISCUSSION THREAD

**Discussion by alexshpilkin** [judofyr/spice#12] [Q&A]
The full binary tree example in the README is good and perhaps representative of some places where one’d want to use automatic parallelization. But there’s also another simple case of a very different flavour: arrays.

Suppose there’s a large linear array of stuff we need to go through in some semblance of an order (a map, a reduce with an associative but not necessarily commutative operation, or something in that vein). Enqueueing every processing job from 0 to N–1 and then gathering up all the results from N–1 to 0 is likely not a good idea for a multitude of reasons (even for a map), one among them being that all the work for enqueueing the items is stuck on the initial thread.

Pretending the array is actually a tree (with the first level dividing the items into [0, N/2) and [N/2, N), and so on down) is probably better, but once the parallelism runs out it has a fairly cache-unfriendly access pattern compared to the linear one we’d have had if we hadn’t went for parallel processing. It sounds like what we’d actually need is to lay out the array non-linearly from the start. I recall a paper ([arXiv][], [ACM DL][]) considered array layouts for in-RAM binary search and measured things like a BFS layout (like you’d use for implementing a binary heap; their winner) and a Van Emde Boas one (theoretical cache-oblivious winner, but in practice a loser for RAM datasets because of the indexing overhead). Their problem is not quite like ours, though.

Thoughts? Experiences?

(I really hope I’m overcomplicating things and there’s a simple solution I overlooked...)

[arXiv]: https://arxiv.org/abs/1509.05053
[ACM DL]: https://dl.acm.org/doi/10.1145/3053370

**Comment by judofyr** [comment]
I agree that "pretending the array is an tree" should be possible, but that will also lead `log(n)` additional stack space which doesn't seem so nice.

I had envisioned something like this:

1. On the stack, we place a task which stores (1) the array and (2) `from: usize`, `to: usize`. We set `from=0, to=N` on the initial one.
2. We start iterating over the array and then we keep incrementing `from` so that it always represents the "not-yet-executed" parts of the array.
3. When the heartbeat kicks we might discover this task. We can then _split_ it: We pick a `k` update the _original_ task to have `to=k` (so that it stops) and then we create a _new_ task with `from=k, to=original.to`.

This kinda creates that tree "on-demand".

However, it's not entire clear to me how the ownership of the data ought to be, nor how we pick a good `k` (or if that matters at all?)

**Reply by alexshpilkin** [reply]
I wanted to complain that the solution you describe is unimplementable on top of the standard fork/join API, but then I realized that in spite of that it might in fact be implementable on top of *your* fork/try-join API, where the try-join call does not take it upon itself to complete a task synchronously if it hasn’t been taken up by another thread. That is an advantage I did not expect.

For example, here’s a version of your idea with _k_ = (_from_ + _to_) / 2, ignoring edge cases:

1. Have _from_ and _to_ passed to us (0 and _N_ respectively in the root task) and introduce _k_ = (_from_ + _to_) / 2.
2. Fork a copy of ourselves passing it _k_ and _to_.
3. Process the first two elements, _from_ and _from_ + 1, and advance _from_ by 2.
4. Try to join the task from step 2.
5. If it has been picked up, do a synchronous recursive call passing it _from_ and _k_, then wait for the task to complete and return.
6. Otherwise, we have withdrawn the task. Advance _k_ by 1 and loop around to step 2.

I think this might be what I wanted. (Now I’ll just figure out why my C port doesn’t scale and then I’ll see if this actually works...)

**Reply by judofyr** [reply]
That sounds possible, yes. Excited to hear how this pans out!

**Reply by theY4Kman** [reply]
FWIW, I implemented something very similar for a mandelbrot rendering loop. Though, I wasn't able to achieve results by only forking off half the work, performing some pixel rendering myself, then trying to join — even if I manually called `t.tick()` at the end of my work.

What ended up working was:
1. Fork the top half of the work
2. Fork a small batch of work from beginning of the set
3. Synchronously perform a batch of work
4. Join on the small batch, performing it ourselves (through `t.call`) if not picked up
5. Join on the top-half batch
	- If picked up, use `t.call` to perform the remainder of our bottom half of the work
	- If not picked up, restore all the work to ourselves, and loop back to step 1

Here's my code (pardon the rough edges; this is the first project I've ever written in Zig). Note that `splitHigh()` and `splitLowBatch()` both mutate the `to` and `from` in `params`.

```zig
fn doRenderParallelWorker(t: *spice.Task, params: *RenderWorkerParams) void {
    const origTo = params.to;

    while (params.from < params.to) {
        var highFut = RenderWorkerFuture.init();
        var highParams: RenderWorkerParams = undefined;

        var hasBatch1 = false;
        var hasBatch1 = false;
        var batch1Fut: RenderWorkerFuture = RenderWorkerFuture.init();
        var batch1Params: RenderWorkerParams = undefined;

        if (params.splitHigh()) |next| {
            highParams = next;
            highFut.fork(t, doRenderParallelWorker, &highParams);
        }

        if (params.splitLowBatch()) |batch| {
            hasBatch1 = true;
            batch1Params = batch;
            batch1Fut.fork(t, doRenderParallelWorker, &batch1Params);
        }

        const escapeScale = mandelbrotZoomScale(params.scale);
        const endIndex = @min(params.to, params.from + RENDER_WORKER_BATCH_SIZE);
        while (params.from < endIndex) : (params.from += 1) {
            const rowCol = params.indexToRowCol(params.from);
            const point = params.projectRowCol(rowCol);
            const escapeCount = t.call(?usize, calculateEscapeCount, point) orelse MAX_ITERATIONS;
            pixels.?.setPixel(rowCol.col, rowCol.row, colorGrad.at(@max(0, escapeCount)));
        }

        if (hasBatch1 and (batch1Fut.tryJoin(t) == null)) {
            t.call(void, doRenderParallelWorker, &batch1Params);
        }

        if (highFut.tryJoin(t)) |_| {
            t.call(void, doRenderParallelWorker, params);
            return;
        } else {
            params.to = origTo;
        }
    }
}
```

[full source here](https://github.com/theY4Kman/experimentation/blob/2d4e083237a59d103c99360a26aa0916f8a18159/graphics/fractals/mandelbrot/zig-mandelbrot/src/main.zig#L221-L263)
